server/src/main/java/org/apache/druid/metadata/input/SqlEntity.java

```diff
@@ -0,0 +1,190 @@
+      boolean foldCase,
+      ObjectMapper objectMapper
+  )
+  {
+    this.sql = sql;
+    this.sqlFirehoseDatabaseConnector = sqlFirehoseDatabaseConnector;
+    this.foldCase = foldCase;
+    this.objectMapper = objectMapper;
+  }
+
+  public String getSql()
+  {
+    return sql;
+  }
+
+  @Nullable
+  @Override
+  public URI getUri()
+  {
+    return null;
+  }
+
+  @Override
+  public InputStream open()
+  {
+    throw new UnsupportedOperationException("Please use fetch() instead");
+  }
+
+  @Override
+  public CleanableFile fetch(File temporaryDirectory, byte[] fetchBuffer) throws IOException
+  {
+    final File tempFile = File.createTempFile("druid-sql-entity", ".tmp", temporaryDirectory);
+    return openCleanableFile(sql, sqlFirehoseDatabaseConnector, objectMapper, foldCase, tempFile);
+
+  }
+
+  public static CleanableFile openCleanableFile(
+      String sql,
+      SQLFirehoseDatabaseConnector sqlFirehoseDatabaseConnector,
+      ObjectMapper objectMapper,
+      boolean foldCase,
+      File tempFile
+  )
+      throws IOException
+  {
+    Preconditions.checkNotNull(sqlFirehoseDatabaseConnector, "SQL Metadata Connector not configured!");
+    try (FileOutputStream fos = new FileOutputStream(tempFile)) {
+      final JsonGenerator jg = objectMapper.getFactory().createGenerator(fos);
+      sqlFirehoseDatabaseConnector.retryWithHandle(
+          (handle) -> {
+            ResultIterator<Map<String, Object>> resultIterator = handle.createQuery(
+                sql
+            ).map(
+                (index, r, ctx) -> {
+                  Map<String, Object> resultRow = foldCase ? new CaseFoldedMap() : new HashMap<>();
+                  ResultSetMetaData resultMetadata;
+                  try {
+                    resultMetadata = r.getMetaData();
+                  }
+                  catch (SQLException e) {
+                    throw new ResultSetException("Unable to obtain metadata from result set", e, ctx);
+                  }
+                  try {
+                    for (int i = 1; i <= resultMetadata.getColumnCount(); i++) {
+                      String key = resultMetadata.getColumnName(i);
+                      String alias = resultMetadata.getColumnLabel(i);
+                      Object value = r.getObject(i);
+                      resultRow.put(alias != null ? alias : key, value);
+                    }
+                  }
+                  catch (SQLException e) {
+                    throw new ResultSetException("Unable to access specific metadata from " +
+                                                 "result set metadata", e, ctx);
+                  }
+                  return resultRow;
+                }
+            ).iterator();
+            jg.writeStartArray();
+            while (resultIterator.hasNext()) {
+              jg.writeObject(resultIterator.next());
+            }
+            jg.writeEndArray();
+            jg.close();
```

Comment:
Are there any size restrictions here? What happens if I try to ingest a very large sql output? How big do the indexing machines need to be? How long can the db connection be kept open so that we can keep writing records to the temp file? Are there any logs that indicate what I need to do operationally to support this?

Reply:
Yeah this is something that I’d like to fix in a follow up PR. There isn’t a way to restrict the file sizes presently  and I feel size limiting on fetched files should be supported at the `InputEntity` interface level. I have added some info in the docs related to the size limit for now.